indexer/validate_and_format_contacts.cpp

```diff
@@ -406,4 +416,97 @@ bool ValidateLinePage(string const & page)
   return (domain == "line.me" || strings::EndsWith(domain, ".line.me"));
 }
 
+bool isSocialContactTag(string_view tag)
+{
+  return tag == "contact:instagram" || tag == "contact:facebook" || tag == "contact:twitter"
+         || tag == "contact:vk" || tag == "contact:line";
+}
+
+bool isSocialContactTag(MapObject::MetadataID const metaID)
+{
+  return metaID == MapObject::MetadataID::FMD_CONTACT_INSTAGRAM ||
+         metaID == MapObject::MetadataID::FMD_CONTACT_FACEBOOK ||
+         metaID == MapObject::MetadataID::FMD_CONTACT_TWITTER ||
+         metaID == MapObject::MetadataID::FMD_CONTACT_VK ||
+         metaID == MapObject::MetadataID::FMD_CONTACT_LINE;
+}
+
+// Functions ValidateAndFormat_{facebook,instagram,twitter,vk}(...) by default strip domain name
+// from OSM data and user input. This function prepends domain name to generate full URL.
+string socialContactToURL(string_view tag, string_view value)
+{
+  if(value.empty())
+    return {};
+
+  if (tag == "contact:instagram")
+    return "https://instagram.com/" + string(value);
+  if (tag == "contact:facebook")
+    return "https://facebook.com/" + string(value);
+  if (tag == "contact:twitter")
+    return "https://twitter.com/" + string(value);
+  if (tag == "contact:vk")
+    return "https://vk.com/" + string(value);
+
+  if (tag == "contact:line")
+  {
+    if (value.find('/') == string::npos) // 'value' is a username.
+      return "https://line.me/R/ti/p/@" + string(value);
+    else // 'value' is an URL.
+      return "https://" + string(value);
+  }
+
+  return string(value);
+}
+
+string socialContactToURL(MapObject::MetadataID metaID, string_view value)
+{
+  if(value.empty())
+    return {};
+
+  switch (metaID)
+  {
+    case MapObject::MetadataID::FMD_CONTACT_INSTAGRAM:
+      return "https://instagram.com/" + string(value);
+    case MapObject::MetadataID::FMD_CONTACT_FACEBOOK:
+      return "https://facebook.com/" + string(value);
+    case MapObject::MetadataID::FMD_CONTACT_TWITTER:
+      return "https://twitter.com/" + string(value);
+    case MapObject::MetadataID::FMD_CONTACT_VK:
+      return "https://vk.com/" + string(value);
+    case MapObject::MetadataID::FMD_CONTACT_LINE:
+      if (value.find('/') == string::npos) // 'value' is a username.
+        return "https://line.me/R/ti/p/@" + string(value);
+      else // 'value' is an URL.
```

Comment:
Isn't Line normalized? Why do we store URLs for it?

Actually, it can be safe to assume that all URLs are normalized in the new release (after adding force-normalization into the generator) because now we don't allow users to edit the map if a new maps data update is available.

Reply:
Line normalization is tricky. You can find some examples in `validate_and_format_contacts_test.cpp` file. E.g.

* `https://line.me/R/ti/p/gdltt7s380` → can extract username `gdltt7s380`
* `https://line.me/R/home/public/profile?id=r90ck7n1rq` → can extract username `r90ck7n1rq`
* `https://liff.line.me/1645278921-kWRPP32q/?accountId=673watcr` → no username and different domain
* `https://page.line.me/?accountId=673watcr` → different domain and `accountId` is not the same thing as username.

So we store domain (no HTTP(S) protocol) for `contact:line`.